modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java

```diff
@@ -1158,33 +1174,75 @@ private static void throwIfSchemaValidationOnCommitFailed(CompatValidationResult
     }
 
     private CompletableFuture<Void> finishAndCleanup(
-            TxFinishReplicaRequest request,
+            Map<TablePartitionId, IgniteBiTuple<ClusterNode, Long>> partitions,
             boolean commit,
-            List<TablePartitionId> aggregatedGroupIds,
+            @Nullable HybridTimestamp commitTimestamp,
             UUID txId,
             String txCoordinatorId
     ) {
-        HybridTimestamp commitTimestamp = request.commitTimestamp();
-
-        CompletableFuture<?> changeStateFuture = finishTransaction(aggregatedGroupIds, txId, commit, commitTimestamp, txCoordinatorId);
-
-        CompletableFuture<?>[] cleanupFutures = new CompletableFuture[request.groups().size()];
-        AtomicInteger cleanupFuturesCnt = new AtomicInteger(0);
-
-        request.groups().forEach(
-                (recipientNode, tablePartitionIds) ->
-                        cleanupFutures[cleanupFuturesCnt.getAndIncrement()] = changeStateFuture.thenCompose(ignored ->
-                                txManager.cleanup(
-                                        recipientNode,
-                                        tablePartitionIds,
-                                        txId,
-                                        commit,
-                                        request.commitTimestamp()
-                                )
-                        )
+        CompletableFuture<?> changeStateFuture = finishTransaction(partitions.keySet(), txId, commit, commitTimestamp, txCoordinatorId);
+
+        CompletableFuture<?>[] futures = partitions.entrySet().stream()
+                .map(entry -> {
+                    TablePartitionId partitionId = entry.getKey();
+                    IgniteBiTuple<ClusterNode, Long> nodeAndTerm = entry.getValue();
+                    return changeStateFuture.thenCompose(ignored ->
+                            cleanupWithRetry(commit, commitTimestamp, txId, partitionId, nodeAndTerm.get1(), nodeAndTerm.get2())
+                    );
+                }).toArray(size -> new CompletableFuture<?>[size]);
+
+        return allOf(futures);
+    }
+
+    private CompletableFuture<Void> cleanupWithRetry(
+            boolean commit,
+            @Nullable HybridTimestamp commitTimestamp,
+            UUID txId,
+            TablePartitionId partitionId,
+            ClusterNode recipientNode,
+            Long term) {
+        CompletableFuture<Void> cleanupFuture = txManager.cleanup(
+                recipientNode,
+                partitionId,
+                term,
+                txId,
+                commit,
+                commitTimestamp
         );
 
-        return allOf(cleanupFutures);
+        if (cleanupFuture == null) {
```

Comment:
How it's possible to have null instead of cleanupFuture. I didn't check the semantics beneath, just curious why it's not possible to have completed future?